jskat-base/src/main/java/org/jskat/control/SkatGame.java

```diff
@@ -134,87 +134,27 @@ public SkatGameResult run() {
 				setGameState(GameState.BIDDING);
 				break;
 			case BIDDING:
-				setActivePlayer(Player.MIDDLEHAND);
-
-				if (variant == GameVariant.FORCED_RAMSCH) {
-					// ramsch games are enforced
-					final GameAnnouncementFactory gaf = GameAnnouncement.getFactory();
-					gaf.setGameType(GameType.RAMSCH);
-					setGameAnnouncement(gaf.getAnnouncement());
-				} else {
-					// "normal" game (i.e. no ramsch)
-					bidding();
-				}
-
-				if (GameType.PASSED_IN.equals(data.getGameType())) {
-					setGameState(GameState.PRELIMINARY_GAME_END);
-				} else if (GameType.RAMSCH.equals(data.getGameType())) {
-					setGameState(GameState.RAMSCH_GRAND_HAND_ANNOUNCING);
-				} else {
-					view.setDeclarer(tableName, data.getDeclarer());
-					setGameState(GameState.PICKING_UP_SKAT);
-				}
+				runBidding();
 				break;
 			case RAMSCH_GRAND_HAND_ANNOUNCING:
-				final boolean grandHandAnnounced = grandHand();
-
-				if (grandHandAnnounced) {
-					log.debug(data.getDeclarer() + " is playing grand hand"); //$NON-NLS-1$
-					final GameAnnouncementFactory gaf = GameAnnouncement.getFactory();
-					gaf.setGameType(GameType.GRAND);
-					gaf.setHand(Boolean.TRUE);
-					setGameAnnouncement(gaf.getAnnouncement());
-					setGameState(GameState.TRICK_PLAYING);
-					log.debug("grand hand game started"); //$NON-NLS-1$
+				if(runRamschGrandHandAnnouncing() == true){
 					break;
```

Comment:
Those left overs don't make any sense. Please just extract the method and turn it into a void method like all the other run methods.

Reply:
I am sorry, I am not sure that it can be a void method. The break inside the if construction should stop the do-while loop.

Reply:
The break in the if block of original code is superfluous. The break at the end of the block should do the trick.

Reply:
Sorry, I do not know.